<a href="https://colab.research.google.com/github/Rudrajit12/Used-Car-Listings-Analysis/blob/main/CarDekho_Data_Cleaning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **🚗 Business Case Study: CarDekho – Feature Importance & Demand Dashboard**

### **🧩 Problem Statement:**

CarDekho, a leading online platform for buying and selling used cars in India, wants to improve its decision-making and customer experience by leveraging the vast data it collects from car listings. With thousands of cars listed daily, the company wants to identify key patterns and trends based on car features, price, and demand, which will then power a centralized dashboard for internal business teams.

This dashboard will help answer key business questions like:

- Which car models and brands are most popular?

- What features (fuel type, transmission, ownership, engine power, etc.) influence car prices the most?

- Which price ranges are in highest demand?

- Are there any underpriced cars with great features (potential for promotion)?

- How do demand patterns differ by seller type (Individual vs Dealer)?

### **✅ Step 1: Problem Understanding**

**🎯 Objective:**

Our goal is to build a data-driven dashboard in Tableau that helps business stakeholders:

- Understand market demand based on car features.

- Identify popular configurations (e.g., manual diesel cars from Maruti with 5 seats).

- Spot high-performing price segments and top-selling car types.

- Enable inventory and pricing strategy improvements for dealers.

We’ll achieve this in two stages:

- Data Cleaning & Preparation in Python.

- Dashboard Development in Tableau.

### **📥 Step 2: Data Collection / Loading the Dataset**

**🎯 Objective:**

Load the dataset into a Pandas DataFrame so we can begin inspecting and preparing it for analysis and visualization.

📝 What We're Doing:

- Import essential libraries.

- Load the dataset from a .csv file.

- Display the first few rows to ensure the data is loaded correctly.

In [ ]:
# Step 1: Import necessary libraries
import pandas as pd
import numpy as np

# Mount the drive
from google.colab import drive
drive.mount('/content/drive/')

# Step 2: Load the dataset
df = pd.read_csv('/content/drive/MyDrive/EDA Projects/CarDekho Car Feature Insights/Car details v3.csv')

# Step 3: View the first 5 rows
print(df.head())

Mounted at /content/drive/
                           name  year  selling_price  km_driven    fuel  \
0        Maruti Swift Dzire VDI  2014         450000     145500  Diesel   
1  Skoda Rapid 1.5 TDI Ambition  2014         370000     120000  Diesel   
2      Honda City 2017-2020 EXi  2006         158000     140000  Petrol   
3     Hyundai i20 Sportz Diesel  2010         225000     127000  Diesel   
4        Maruti Swift VXI BSIII  2007         130000     120000  Petrol   

  seller_type transmission         owner     mileage   engine   max_power  \
0  Individual       Manual   First Owner   23.4 kmpl  1248 CC      74 bhp   
1  Individual       Manual  Second Owner  21.14 kmpl  1498 CC  103.52 bhp   
2  Individual       Manual   Third Owner   17.7 kmpl  1497 CC      78 bhp   
3  Individual       Manual   First Owner   23.0 kmpl  1396 CC      90 bhp   
4  Individual       Manual   First Owner   16.1 kmpl  1298 CC    88.2 bhp   

                     torque  seats  
0            190Nm@ 20

### **🔎 Step 3: Data Inspection**

**🎯 Objective:**

Understand the structure, data types, and basic statistics of the dataset to identify potential issues or areas that need cleaning or transformation.

📝 What We're Doing:

We will:

- Check the first few rows to understand the data format.

- Use .info() to check column types and missing values.

- Use .describe() to get statistical summaries of numeric columns.

- Check how many unique values each column has (for categorical assessment).

- Use .nunique() and .columns to get an overview of features.

In [ ]:
# Check the shape of the dataset (rows, columns)
print("Shape of dataset:", df.shape)

Shape of dataset: (8128, 13)


In [ ]:
# Get detailed info: data types, non-null counts
print("\nData Types and Null Counts:")
print(df.info())


Data Types and Null Counts:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8128 entries, 0 to 8127
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   name           8128 non-null   object 
 1   year           8128 non-null   int64  
 2   selling_price  8128 non-null   int64  
 3   km_driven      8128 non-null   int64  
 4   fuel           8128 non-null   object 
 5   seller_type    8128 non-null   object 
 6   transmission   8128 non-null   object 
 7   owner          8128 non-null   object 
 8   mileage        7907 non-null   object 
 9   engine         7907 non-null   object 
 10  max_power      7913 non-null   object 
 11  torque         7906 non-null   object 
 12  seats          7907 non-null   float64
dtypes: float64(1), int64(3), object(9)
memory usage: 825.6+ KB
None


In [ ]:
# Summary statistics for numerical columns
print("\nStatistical Summary (Numerical):")
print(df.describe())


Statistical Summary (Numerical):
              year  selling_price     km_driven        seats
count  8128.000000   8.128000e+03  8.128000e+03  7907.000000
mean   2013.804011   6.382718e+05  6.981951e+04     5.416719
std       4.044249   8.062534e+05  5.655055e+04     0.959588
min    1983.000000   2.999900e+04  1.000000e+00     2.000000
25%    2011.000000   2.549990e+05  3.500000e+04     5.000000
50%    2015.000000   4.500000e+05  6.000000e+04     5.000000
75%    2017.000000   6.750000e+05  9.800000e+04     5.000000
max    2020.000000   1.000000e+07  2.360457e+06    14.000000


In [ ]:
# Summary statistics for object (categorical) columns
print("\nSummary of Categorical Columns:")
print(df.describe(include='object'))


Summary of Categorical Columns:
                          name    fuel seller_type transmission        owner  \
count                     8128    8128        8128         8128         8128   
unique                    2058       4           3            2            5   
top     Maruti Swift Dzire VDI  Diesel  Individual       Manual  First Owner   
freq                       129    4402        6766         7078         5289   

          mileage   engine max_power          torque  
count        7907     7907      7913            7906  
unique        393      121       322             441  
top     18.9 kmpl  1248 CC    74 bhp  190Nm@ 2000rpm  
freq          225     1017       377             530  


In [ ]:
# Check number of unique values in each column
print("\nUnique values per column:")
print(df.nunique())

# Display all column names
print("\nColumn Names:")
print(df.columns.tolist())


Unique values per column:
name             2058
year               29
selling_price     677
km_driven         921
fuel                4
seller_type         3
transmission        2
owner               5
mileage           393
engine            121
max_power         322
torque            441
seats               9
dtype: int64

Column Names:
['name', 'year', 'selling_price', 'km_driven', 'fuel', 'seller_type', 'transmission', 'owner', 'mileage', 'engine', 'max_power', 'torque', 'seats']


After loading the dataset, we conducted an initial inspection to understand its structure, data types, and completeness. Below are the key findings:

**📦 Dataset Overview:**

- Total Records: 8128 rows

- Total Features: 13 columns

The dataset includes a mix of categorical, numerical, and text-based features representing car specifications, seller details, and performance metrics.

---

**📊 Data Type Summary:**

- Numerical Columns (int/float):

  - year, selling_price, km_driven – all correctly typed as integers.

  - seats – stored as float due to presence of missing values.

- Categorical/Text Columns (object):

  - name, fuel, seller_type, transmission, owner – clean object types.

  - mileage, engine, max_power, torque – stored as object, even though they contain numeric-like values mixed with units (e.g., "23.4 kmpl", "1248 CC").

---

**⚠️ Missing Value Summary:**

Several columns have missing values, especially those representing car performance:

- mileage: 221 missing

- engine: 221 missing

- max_power: 215 missing

- torque: 222 missing

- seats: 221 missing

The remaining columns are complete with no null values, including key identifiers and categorical variables.

---

**🔎 Early Observations:**

- The dataset is largely well-structured, but certain columns like mileage, engine, and max_power require cleaning to extract numeric values from strings.

- The torque column appears to have complex, non-uniform textual data, which may not be directly usable without additional parsing or feature engineering.

- The seats column, while numeric, contains missing values that should be addressed through imputation or analysis of patterns.

A descriptive analysis of the numerical columns reveals several important patterns and outliers that will be valuable for data cleaning and dashboard insights.

**🚗 Year (Manufacturing Year):**

- Range: 1983 to 2020

- Median (50%): 2015

- Observation: Most cars were manufactured between 2011 and 2017.

- Outlier Alert: Cars older than 2000 may be rare listings and require review for relevance.

---

**💰 Selling Price:**

- Range: ₹29,999 to ₹1 crore

- Mean: ₹6.38 lakhs

- Median: ₹4.5 lakhs

- Observation: The price distribution is right-skewed, with some luxury car listings significantly pulling up the average.

- Business Insight: Majority of listings fall between ₹2.5 to ₹6.75 lakhs, suggesting that affordable used cars dominate the market.

---

**📉 Kilometers Driven:**

- Range: 1 km to over 23 lakh km (2,360,457 km)

- Median: 60,000 km

- Observation: The spread is wide, but most used cars fall within 35,000 to 98,000 km driven.

- Outlier Alert: Some extremely high values (over 2 million km) could be data entry errors or fleet vehicles.

---

**🪑 Seats:**

- Range: 2 to 14 seats

- Mean: ~5.42

- Median: 5 seats

- Observation: The vast majority of cars are 5-seaters, which aligns with the dominance of hatchbacks and sedans in the Indian used car market.

- Outlier Alert: Listings with 9–14 seats likely represent commercial vehicles or vans.

---

**🧠 Summary of Insights:**

- The market is centered around mid-range pricing and mileage, making affordability a key focus.

- 5-seater cars from 2011–2017 dominate the listings — crucial for understanding demand trends.

- Some extreme values (price, km_driven, seats) require capping or outlier treatment before visualization or modeling.

### **🧹 Step 4: Data Cleaning**

**🎯 Objective:**

Clean the dataset by addressing:

- Missing values

- Improper data types

- Textual inconsistencies

- Outliers or unusable columns

📝 What We're Doing:

We'll perform the following operations:

- Remove duplicate records (if any).

- Clean & convert columns like mileage, engine, max_power from string to float by removing units.

- Impute or drop missing values appropriately.

- Drop unstructured or non-critical columns (torque).

- Check and fix outliers or extreme values where necessary.



In [ ]:
import numpy as np

# Step 1: Drop duplicate rows
df.drop_duplicates(inplace=True)

# Step 2: Convert columns to string before string operations
df['mileage'] = df['mileage'].astype(str).str.replace(' kmpl', '', regex=False).str.replace(' km/kg', '', regex=False)
df['engine'] = df['engine'].astype(str).str.replace(' CC', '', regex=False)
df['max_power'] = df['max_power'].astype(str).str.replace(' bhp', '', regex=False)

# Step 3: Replace empty strings with NaN
df['mileage'].replace(['', 'nan'], np.nan, inplace=True)
df['engine'].replace(['', 'nan'], np.nan, inplace=True)
df['max_power'].replace(['', 'nan'], np.nan, inplace=True)

# Step 4: Convert cleaned columns to float
df['mileage'] = df['mileage'].astype(float)
df['engine'] = df['engine'].astype(float)
df['max_power'] = df['max_power'].astype(float)

# Step 5: Fill missing values for 'seats' with mode
df['seats'].fillna(df['seats'].mode()[0], inplace=True)

# Step 6: Drop rows where critical fields are still missing
df.dropna(subset=['mileage', 'engine', 'max_power'], inplace=True)

# Step 7: Drop 'torque' due to inconsistent formatting
df.drop(columns=['torque'], inplace=True)

# Step 8: Final check
print("✅ Cleaning Complete – Remaining Missing Values:\n")
print(df.isnull().sum())

✅ Cleaning Complete – Remaining Missing Values:

name             0
year             0
selling_price    0
km_driven        0
fuel             0
seller_type      0
transmission     0
owner            0
mileage          0
engine           0
max_power        0
seats            0
dtype: int64


/tmp/ipython-input-9-41719529.py:12: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['mileage'].replace(['', 'nan'], np.nan, inplace=True)
/tmp/ipython-input-9-41719529.py:13: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True

In [ ]:
# Print the data
df.head()

name  year  selling_price  km_driven    fuel  \
0        Maruti Swift Dzire VDI  2014         450000     145500  Diesel   
1  Skoda Rapid 1.5 TDI Ambition  2014         370000     120000  Diesel   
2      Honda City 2017-2020 EXi  2006         158000     140000  Petrol   
3     Hyundai i20 Sportz Diesel  2010         225000     127000  Diesel   
4        Maruti Swift VXI BSIII  2007         130000     120000  Petrol   

  seller_type transmission         owner  mileage  engine  max_power  seats  
0  Individual       Manual   First Owner    23.40  1248.0      74.00    5.0  
1  Individual       Manual  Second Owner    21.14  1498.0     103.52    5.0  
2  Individual       Manual   Third Owner    17.70  1497.0      78.00    5.0  
3  Individual       Manual   First Owner    23.00  1396.0      90.00    5.0  
4  Individual       Manual   First Owner    16.10  1298.0      88.20    5.0

### **🔧 Step 5 (Part 1): Rename Columns for Clarity**

**🎯 Objective:**

Rename technical columns to more descriptive forms — this improves readability in Tableau and helps stakeholders instantly understand what the metric represents.

In [ ]:
# Rename columns for better readability
df.rename(columns={
    'mileage': 'mileage (km/l)',
    'engine': 'engine (cc)',
    'max_power': 'max_power (bhp)'
}, inplace=True)

# Print the data
df.head()

name  year  selling_price  km_driven    fuel  \
0        Maruti Swift Dzire VDI  2014         450000     145500  Diesel   
1  Skoda Rapid 1.5 TDI Ambition  2014         370000     120000  Diesel   
2      Honda City 2017-2020 EXi  2006         158000     140000  Petrol   
3     Hyundai i20 Sportz Diesel  2010         225000     127000  Diesel   
4        Maruti Swift VXI BSIII  2007         130000     120000  Petrol   

  seller_type transmission         owner  mileage (km/l)  engine (cc)  \
0  Individual       Manual   First Owner           23.40       1248.0   
1  Individual       Manual  Second Owner           21.14       1498.0   
2  Individual       Manual   Third Owner           17.70       1497.0   
3  Individual       Manual   First Owner           23.00       1396.0   
4  Individual       Manual   First Owner           16.10       1298.0   

   max_power (bhp)  seats  
0            74.00    5.0  
1           103.52    5.0  
2            78.00    5.0  
3            90.00    5.0  
4            88.20    5.0

### **🔧 Step 5 (Part 2): Clean & Standardize Owner Field**

**🎯 Objective:**

The owner column has verbose entries like "First Owner", "Second Owner", etc. We'll simplify them to:

- "First", "Second", "Third", "Fourth & Above", "Test Drive" (if present)

This standardization helps with filtering and grouping in Tableau.

In [ ]:
# Map owner labels to simpler form
df['owner'] = df['owner'].replace({
    'First Owner': 'First',
    'Second Owner': 'Second',
    'Third Owner': 'Third',
    'Fourth & Above Owner': 'Fourth & Above',
    'Test Drive Car': 'Test Drive'
})

# Print the data
df.head()

name  year  selling_price  km_driven    fuel  \
0        Maruti Swift Dzire VDI  2014         450000     145500  Diesel   
1  Skoda Rapid 1.5 TDI Ambition  2014         370000     120000  Diesel   
2      Honda City 2017-2020 EXi  2006         158000     140000  Petrol   
3     Hyundai i20 Sportz Diesel  2010         225000     127000  Diesel   
4        Maruti Swift VXI BSIII  2007         130000     120000  Petrol   

  seller_type transmission   owner  mileage (km/l)  engine (cc)  \
0  Individual       Manual   First           23.40       1248.0   
1  Individual       Manual  Second           21.14       1498.0   
2  Individual       Manual   Third           17.70       1497.0   
3  Individual       Manual   First           23.00       1396.0   
4  Individual       Manual   First           16.10       1298.0   

   max_power (bhp)  seats  
0            74.00    5.0  
1           103.52    5.0  
2            78.00    5.0  
3            90.00    5.0  
4            88.20    5.0

### **🔧 Step 5 (Part 3): Extract brand_name and model_name from name**

**🎯 Objective:**

Split the name column into:

- brand_name: The first word (e.g., Maruti, Hyundai, Honda)

- model_name: The rest of the name (e.g., Swift Dzire VDI, City i-VTEC VX)

In [ ]:
# Split the 'name' column into brand and model
df['brand_name'] = df['name'].str.split().str[0]  # First word = Brand
df['model_name'] = df['name'].str.split().str[1:].apply(lambda x: ' '.join(x))  # Remaining words = Model

# Print the data
df.head()

name  year  selling_price  km_driven    fuel  \
0        Maruti Swift Dzire VDI  2014         450000     145500  Diesel   
1  Skoda Rapid 1.5 TDI Ambition  2014         370000     120000  Diesel   
2      Honda City 2017-2020 EXi  2006         158000     140000  Petrol   
3     Hyundai i20 Sportz Diesel  2010         225000     127000  Diesel   
4        Maruti Swift VXI BSIII  2007         130000     120000  Petrol   

  seller_type transmission   owner  mileage (km/l)  engine (cc)  \
0  Individual       Manual   First           23.40       1248.0   
1  Individual       Manual  Second           21.14       1498.0   
2  Individual       Manual   Third           17.70       1497.0   
3  Individual       Manual   First           23.00       1396.0   
4  Individual       Manual   First           16.10       1298.0   

   max_power (bhp)  seats brand_name              model_name  
0            74.00    5.0     Maruti         Swift Dzire VDI  
1           103.52    5.0      Skoda  Rapid 1.5 TDI Ambition  
2            78.00    5.0      Honda      City 2017-2020 EXi  
3            90.00    5.0    Hyundai       i20 Sportz Diesel  
4            88.20    5.0     Maruti         Swift VXI BSIII

### **💾 Final Step: Export Cleaned Dataset for Tableau**

In [ ]:
# Export the cleaned dataset to a new CSV file
df.to_csv('/content/drive/MyDrive/EDA Projects/CarDekho Car Feature Insights/cleaned_car_data.csv', index=False)
print("✅ Cleaned data exported successfully to 'cleaned_car_data.csv'")

✅ Cleaned data exported successfully to 'cleaned_car_data.csv'


### **✅ What We Did — Step-by-Step Summary:**

**🔍 Step 1: Problem Understanding**

- Framed the business use-case for CarDekho: identify which car features affect demand and pricing, and build a Tableau dashboard to support decisions.

**📥 Step 2: Data Loading**

- Loaded the dataset using Pandas and began exploring its structure.

**📊 Step 3: Data Inspection**

- Identified data types, null values, and format issues.

- Found that columns like mileage, engine, and max_power were stored as strings with units.

**🧹 Step 4: Data Cleaning**

- Removed duplicates and handled missing values.

- Cleaned and converted:

  - mileage → mileage (km/l)

  - engine → engine (cc)

  - max_power → max_power (bhp)

- Dropped the torque column due to inconsistent text formatting.

- Filled missing values in seats using mode.

**🔧 Step 5: Data Transformation**

- Renamed technical columns for clarity in Tableau.

- Cleaned and simplified owner values (e.g., "First Owner" → "First").

- Extracted:

  - brand_name from the first word in name

  - model_name from the remaining part of the name